### ISSUE TITLE
Reproducing results on LEVIR-CD+

### ISSUE THREAD

**Issue by robmarkcole** [AndreaCodegoni/Tiny_model_4_CD#17]
I'm attempting to reproduce the paper results on LEVIR-CD+ & I note you trained for 100 epochs and apparently didn't use early stopping. Based on my training run it appears learning plateaus after 30-40 epochs, is this what you saw?

<img width="1805" alt="image" src="https://github.com/AndreaCodegoni/Tiny_model_4_CD/assets/11855322/33e1e45b-7c5f-4925-94ca-c2163ddf7f40">

<img width="1192" alt="image" src="https://github.com/AndreaCodegoni/Tiny_model_4_CD/assets/11855322/064b97b6-cbdf-4bfa-94e9-cbe8a0501bd0">

At 100 epochs the metrics are short of those in the paper:
```
 'test_f1': 0.75
 'test_iou': 0.60
```

Possible diff in the dataset used in the paper vs [this version](https://torchgeo.readthedocs.io/en/stable/api/datasets.html#levir-cd)?

**Comment by AndreaCodegoni** [comment]
Hello @robmarkcole.

As you pointed out the dataset I have used in my experiments was LEVIR-CD, which has 637 image pairs. The LEVIR-CD+ is a different dataset and it has 985 image pairs. 
Considering the difference in cardinality in the two datasets I would tell you that I expect fewer epochs to be needed for the model to converge and avoid overfitting. In fact, I seem to see that the metrics are good after 30/40 epochs. I have never experimented on LEVIR-CD+, but you can probably tune the optimizer parameters for better performance.

I see something interesting from the train_loss graph: there is a spike which could mean there has been some catastrophic forgetting of pre-trained weights on ImageNet. It might be interesting to train without ImageNet weights to see if this phenomenon disappears and if you can still achieve the same performance with a sufficient number of epochs. If you give it a try let me know how it goes :)

**Comment by robmarkcole** [comment]
@AndreaCodegoni at your suggestion I have run training with 

```
    pretrained=False,
    freeze_backbone=False,
```

The logs look similar:

<img width="1799" alt="image" src="https://github.com/AndreaCodegoni/Tiny_model_4_CD/assets/11855322/9d11d161-1abb-4e04-a187-180f2c610768">

<img width="1800" alt="image" src="https://github.com/AndreaCodegoni/Tiny_model_4_CD/assets/11855322/aac20d09-6be9-43b1-8ee5-13e70ad92845">


And test metrics still poor:
```
  'test_f1': 0.6987901329994202,
  'test_iou': 0.5370310544967651}]
```

The model rarely predicts the change class:
<img width="1435" alt="image" src="https://github.com/AndreaCodegoni/Tiny_model_4_CD/assets/11855322/bc12656b-3747-4936-a6a3-2d1a2e3297da">

I've noticed that my normalisation is slightly different, rather than bandwise norm you have performed I am normalising 0-1 range. However as not using imagenet weights this shouldn't be an issue?

Is TinyCD particularly sensitive to learning rate etc?

**Comment by AndreaCodegoni** [comment]
Given the results without pre-training, I would go back to using ImageNet weights and try using ImageNet normalization for mean and variance as well (and not just rescaling between 0 and 1).

As for hyperparameters such as learning rate etc. the answer is yes. As also reported in the paper and in the code, optimizing these parameters led to improvements. In the original work they actually improved by a few points, but if you have the opportunity I recommend you try.

**Comment by robmarkcole** [comment]
Having validated my training code in https://github.com/AndreaCodegoni/Tiny_model_4_CD/issues/18 and now performing imagenet normalisation I have re-run the training for 100 epochs. The metrics are slightly worst this time at 
```
  'test_f1': 0.705
  'test_iou': 0.544
```

<img width="1800" alt="image" src="https://github.com/AndreaCodegoni/Tiny_model_4_CD/assets/11855322/77dabfaa-a71e-4a05-9d68-7f4dbdb3e983">

Some predictions are encouraging
<img width="1435" alt="image" src="https://github.com/AndreaCodegoni/Tiny_model_4_CD/assets/11855322/68d28860-b80a-43d8-804e-ad6543dd1c0b">

Since this is a larger dataset it may require more epochs, as well as hyperparam optimisation as you suggest

**Comment by AndreaCodegoni** [comment]
Hi @robmarkcole, sorry for the late reply.

I seem to see that the loss in training is still quite unstable.
Probably on larger datasets TinyCD needs to be trained with some more care as you are suggesting.

I don't know what the distribution of images is like in the dataset, i.e. how many pairs with changes there are vs pairs without changes, but that could be a factor too. Maybe you can try training with a larger batch size or even train batches with a certain ratio between images with and without changes.
